Merge/model db upload val

Leon-Leyang opened the pull request:
## Issue ticket number and link

Tag relevant issues and PRs as follows

- Resolved issue #177, #178 and #183.

## What are the changes

Merged the changes from chonghan/model-db and leyang/model-upload-val.

## Comments

- The `UploadModel` API has been tested manually. But there are still some problems passing CircleCI that need further checking.

## Checklist before requesting a review

- [ ] Double check the diff between my branch and the target branch. Make sure there's nothing unexpected.
- [ ] Tag relevant issues and PRs in the first section of this form.
- [ ] All CI tests are passed.
- [ ] No `package-lock.json` or `package.json` changes are pushed. If you do, add a new section `Dependency Changes` to the form stating what library changes are introduced and why.

PaulCCCCCCH committed:
Add db setup doc; remove useless db files

PaulCCCCCCH committed:
add database schema change note

PaulCCCCCCH committed:
add db schema design diagram and code

Leon-Leyang committed:
Create 'model.py' for initial model APIs sketch

Leon-Leyang committed:
Add docstrings for the UploadModel API

Leon-Leyang committed:
Add 'model_utils.py' for utilities & outline 'UploadModel' API

Leon-Leyang committed:
Get all metadata of the model according to the db design

Leon-Leyang committed:
Complete `init_model` function implementation

Leon-Leyang committed:
Extend `UploadModel` to handle uploading models after training

PaulCCCCCCH committed:
refactor: implemented orm models

PaulCCCCCCH committed:
fixed db model definition bugs

PaulCCCCCCH committed:
change db types in function signatures

Leon-Leyang committed:
Enhance UploadModel API with ID assignment and architecture metadata

Leon-Leyang committed:
Complete `val_model` function implementation

PaulCCCCCCH committed:
fix: new database operations with SQLAlchemy partially working

PaulCCCCCCH committed:
most functionalities working; not passing test cases

PaulCCCCCCH committed:
test cases passing

PaulCCCCCCH committed:
chore: update docker image version; include SQLAlchemy dependency

PaulCCCCCCH committed:
feat: implemented model operations

Leon-Leyang committed:
Update save path for model definition and checkpoint files to /generated/models

Leon-Leyang committed:
Refactor 'UploadModel' API for enhanced model handling

- Improve API to handle predefined models.
- Remove logic handling post-training model uploads.
- Utilize `create_model` from `RModelWrapper` for saving.

Leon-Leyang committed:
Migrate from importlib_resources to importlib.resources

Leon-Leyang committed:
Refresh requirements.txt with updated packge versions

Leon-Leyang committed:
Merge branch 'dev' of https://github.com/HaohanWang/Robustar_implementation into dev

Leon-Leyang committed:
Merge branch 'chonghan/model-db' into 'merge/model-db-upload-val'

Leon-Leyang committed:
Merge branch 'leyang/model-upload-val' into merge/model-db-upload-val

Leon-Leyang committed:
Correct some mistakes to let 'UploadModel' API work for basic custom model upload

- Change `Models` table id type from BigInteger to Integer for autoincrement
- Correct typos

Leon-Leyang committed:
Fix type errors of var `pretrained` and `num_classes`

Leon-Leyang committed:
Save the weight of pretrained predefined model to a local file

Leon-Leyang committed:
Raise an exception if pretrained weight is required for "ResNet18-32x32"

Leon-Leyang committed:
Save `num_classes` of predefined models into local files for later restoration

Leon-Leyang committed:
Save current model weights to a local file if the weight file is not provided

Leon-Leyang committed:
Comment db clean up temporarily.

Leon-Leyang committed:
Permit optional transmission of description and tags in metadata from frontend

Leon-Leyang committed:
Update model metadata design: Introduce `nickname` and Rename `name` to `class_name`

Leon-Leyang committed:
Extract 'tags' column from 'Models' table into a new table, and maintain the relationship

Leon-Leyang committed:
Enhance `UploadModel` docstring documentation

Leon-Leyang committed:
Close DB connection explicitly for .db file removal; Set fixture scope to 'function' for fresh instances per test.

Leon-Leyang committed:
Change the base directory in pytest for debugging

Leon-Leyang committed:
Debugging

Leon-Leyang committed:
Debugging

Leon-Leyang committed:
Debugging

Leon-Leyang committed:
Debugging

Leon-Leyang committed:
Debugging

Leon-Leyang committed:
Debugging

Leon-Leyang committed:
Restore for a full test

Leon-Leyang committed:
Restore for full test

Leon-Leyang committed:
Capture the 'stdout' and 'stderr' in pytest for debugging

Leon-Leyang committed:
Debugging

Leon-Leyang committed:
Debugging

Leon-Leyang committed:
Debugging

Leon-Leyang committed:
Solve the problem by specifying the db file address as absolute path

Leon-Leyang committed:
Restore the config of CircleCI test

PaulCCCCCCH committed:
fix: provide app context to all thread creations

PaulCCCCCCH committed:
Merge branch 'merge/model-db-upload-val' of github.com:HaohanWang/Robustar_migration into merge/model-db-upload-val

PaulCCCCCCH committed:
Merge branch 'dev' into merge/model-db-upload-val

PaulCCCCCCH committed:
fix: correctly getting context for child threads

PaulCCCCCCH commented on back-end/apis/model.py:
```diff
@@ -0,0 +1,270 @@
+        except Exception as e:
+            clear_model_temp_files(saving_id)
+            return RResponse.abort(400, f"Failed to initialize the custom model. {e}")
+    else:   # If the model is predefined
```
This may make it more robust:
```
if 'code' in request.form and request.form['code']:
    ...
elif 'pretrained' in request.form:
    ...
else:
     abort(bad request)
```

PaulCCCCCCH commented on back-end/database/db_ops.py:
```diff
@@ -1,35 +1,38 @@
 from typing import List, Tuple
```
We should be using ORM objects everywhere instead of these SQL templates. I believe this file is no longer needed and we should probably remove it. Could you also double check?

PaulCCCCCCH commented on back-end/utils/model_utils.py:
```diff
@@ -0,0 +1,106 @@
+
+
+# TODO: Use real model wrapper instead of dummy model wrapper
+class DummyModelWrapper:
```
Thanks for creating this! I can merge it into my model switch implementation.

PaulCCCCCCH reviewed (approved):
Thanks. Left some comment.

PaulCCCCCCH commented:
Feel free to convert this draft PR to a real one and merge once you are done with the updates and pass all existing tests again

Leon-Leyang committed:
Resolve code review feedback:

- Improve robustness by checking for 'code' and 'pretrained' in request
- Remove 'back-end/database/db_ops.py' file (double-checked for necessity)

Closes: #192

Leon-Leyang committed:
Merge remote-tracking branch 'origin/merge/model-db-upload-val' into merge/model-db-upload-val
